lib/std/crypto/rsa.zig

```diff
@@ -0,0 +1,873 @@
+    /// `d`
+    private_exponent: Fe,
+
+    pub const FromBytesError = ValueError || ff.OverflowError || ff.FieldElementError;
+
+    pub fn fromBytes(n: Modulus, exp: []const u8) FromBytesError!SecretKey {
+        const d = try Fe.fromBytes(n, exp, .big);
+        if (std.debug.runtime_safety) {
+            // > The RSA private exponent d is a positive integer less than n
+            // > satisfying e * d == 1 (mod \lambda(n)),
+            if (!d.isOdd()) return error.Exponent;
+            if (d.v.compare(n.v) != .lt) return error.Exponent;
+        }
+
+        return .{ .private_exponent = d };
+    }
+};
+
+pub const KeyPair = struct {
+    public: PublicKey,
+    secret: SecretKey,
+
+    pub const FromDerError = PublicKey.FromBytesError || SecretKey.FromBytesError || der.Parser.Error || error{ KeyMismatch, InvalidVersion };
+
+    pub fn fromDer(bytes: []const u8) FromDerError!KeyPair {
+        var parser = der.Parser{ .bytes = bytes };
+        const seq = try parser.expectSequence();
+        const version = try parser.expectInt(u8);
+
+        const mod = try parser.expectPrimitive(.integer);
+        const pub_exp = try parser.expectPrimitive(.integer);
+        const sec_exp = try parser.expectPrimitive(.integer);
+
+        const public = try PublicKey.fromBytes(parser.view(mod), parser.view(pub_exp));
+        const secret = try SecretKey.fromBytes(public.modulus, parser.view(sec_exp));
+
+        const prime1 = try parser.expectPrimitive(.integer);
+        const prime2 = try parser.expectPrimitive(.integer);
+        const exp1 = try parser.expectPrimitive(.integer);
+        const exp2 = try parser.expectPrimitive(.integer);
+        const coeff = try parser.expectPrimitive(.integer);
+        _ = .{ exp1, exp2, coeff };
+
+        switch (version) {
+            0 => {},
+            1 => {
+                _ = try parser.expectSequenceOf();
+                while (!parser.eof()) {
+                    _ = try parser.expectSequence();
+                    const ri = try parser.expectPrimitive(.integer);
+                    const di = try parser.expectPrimitive(.integer);
+                    const ti = try parser.expectPrimitive(.integer);
+                    _ = .{ ri, di, ti };
+                }
+            },
+            else => return error.InvalidVersion,
+        }
+
+        try parser.expectEnd(seq.slice.end);
+        try parser.expectEnd(bytes.len);
+
+        if (std.debug.runtime_safety) {
+            const p = try Fe.fromBytes(public.modulus, parser.view(prime1), .big);
+            const q = try Fe.fromBytes(public.modulus, parser.view(prime2), .big);
+
+            // check that n = p * q
+            const expected_zero = public.modulus.mul(p, q);
+            if (!expected_zero.isZero()) return error.KeyMismatch;
+
+            // TODO: check that d * e is one mod p-1 and mod q-1. Note d and e were bound
+            // const de = secret.private_exponent.mul(public.public_exponent);
```

Comment:
Need some help here. Can also just remove this incorrect commented code for now.